google/resource_dataproc_cluster.go

```diff
@@ -0,0 +1,870 @@
+			m[k] = val.(string)
+		}
+		cluster.Config.GceClusterConfig.Metadata = m
+	}
+
+	// Create the cluster
+	op, err := config.clientDataproc.Projects.Regions.Clusters.Create(
+		project, region, cluster).Do()
+	if err != nil {
+		return err
+	}
+
+	// Wait until it's created
+	waitErr := dataprocClusterOperationWait(config, op, "creating Dataproc cluster", timeoutInMinutes, 3)
+	if waitErr != nil {
+		// The resource didn't actually create
+		d.SetId("")
+		return waitErr
+	}
+
+	log.Printf("[INFO] Dataproc cluster %s has been created", clusterName)
+	d.SetId(clusterName)
+
+	e := resourceDataprocClusterRead(d, meta)
+	if e != nil {
+		log.Print("[INFO] Got an error reading back after creating, \n", e)
+	}
+	return e
+}
+
+func resourceDataprocClusterUpdate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	// The only items which are currently able to be updated, without a
+	// forceNew in place are the labels and/or the number of worker nodes in a cluster
+	if !(d.HasChange("labels") ||
+		d.HasChange("worker_config.0.num_workers") ||
+		d.HasChange("worker_config.0.preemptible_num_workers")) {
+		return errors.New("*** programmer issue - update resource called however item is not allowed to be changed - investigate ***")
+	}
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	region := d.Get("region").(string)
+	clusterName := d.Get("name").(string)
+	timeoutInMinutes := int(d.Timeout(schema.TimeoutUpdate).Minutes())
+
+	cluster := &dataproc.Cluster{
+		ClusterName: clusterName,
+		ProjectId:   project,
+		Config:      &dataproc.ClusterConfig{},
+	}
+	patch := config.clientDataproc.Projects.Regions.Clusters.Patch(
+		project, region, clusterName, cluster)
+
+	updMask := ""
+
+	if d.HasChange("labels") {
+
+		v := d.Get("labels")
+		m := make(map[string]string)
+		for k, val := range v.(map[string]interface{}) {
+			m[k] = val.(string)
+		}
+		cluster.Labels = m
+		updMask = "labels"
+	}
+
+	if d.HasChange("worker_config.0.num_workers") {
+
+		wconfigs := d.Get("worker_config").([]interface{})
+		conf := wconfigs[0].(map[string]interface{})
+
+		desiredNumWorks := conf["num_workers"].(int)
+		cluster.Config.WorkerConfig = &dataproc.InstanceGroupConfig{
+			NumInstances: int64(desiredNumWorks),
+		}
+
+		if len(updMask) > 0 {
+			updMask = updMask + ","
+		}
+		updMask = updMask + "config.worker_config.num_instances"
+	}
+
+	if d.HasChange("worker_config.0.preemptible_num_workers") {
+
+		wconfigs := d.Get("worker_config").([]interface{})
+		conf := wconfigs[0].(map[string]interface{})
+
+		desiredNumWorks := conf["preemptible_num_workers"].(int)
+		cluster.Config.SecondaryWorkerConfig = &dataproc.InstanceGroupConfig{
+			NumInstances: int64(desiredNumWorks),
+		}
+
+		if len(updMask) > 0 {
+			updMask = updMask + ","
+		}
+		updMask = updMask + "config.secondary_worker_config.num_instances"
+	}
+
+	patch.UpdateMask(updMask)
+
+	op, err := patch.Do()
```

Comment:
This is slightly different as I need to set the UpdateMask which cannot be set in the constructor as with the others. If you chain it it looks a bit messy so did at least move it down so its all together.